### PR TITLE
Fix carousel not returning to previous selection after filter

### LINKED ISSUE #2638: Filtering beatmaps after null selection
While fixing #2264 I never looked at what stable does. In stable the selection never goes to null, as in there's always some beatmap playing/displayed in the background, which is good in one situation.

A user might mistype while trying to further filter their selection, which in lazer would result in you losing your index (for a lack of better word) after being placed to random spot after your now correct filter. In stable the selection would just go back to the previous beatmap that was playing before the filter matched no beatmaps.

Example:
* You first search ``touhou``
* Go through half your library and then want to further refine the search term
* You now search for ``touhou imperishable`` but mistype imperishable
* In stable you are greeted with ``No matches found. Hit ESC to reset.`` and after correcting your mistake you can continue browsing your library from where you left off.
* In lazer you are greeted with ``no beatmaps avalibe! please load a beatmap!`` and after correcting your mistake are sent to random spot in your library and you have to find the spot you were at before. 

I propose that instead of random selection, last valid beatmap is used, unless that beatmap is now filtered in which case keep the current behaviour. Last valid beatmap could be just the last one, could be last 50 ones and the algorithm chooses the latest.

### PR DESCRIPTION
**Pull request by peppy**
Alternative to and closes https://github.com/ppy/osu/pull/8251.

- Closes #2638.

This logic already existed but wasn't being executed due to `LastSelected` getting incorrectly set to null. Basically a one line change (although I had to make one more change to keep correct test behaviour).

### CODE REVIEW

**Comment by LittleEndu** [comment]
If this also addresses #8239

**Comment by peppy** [comment]
This does not address #8239. Only #2638 I guess.

**Comment by peppy** [comment]
Actually it looks to solve both.

**Comment by peppy** [comment]
No, it definitely doesn't. #8239 requires a different solution.

File: osu.Game/Screens/Select/BeatmapCarousel.cs
Comment (LittleEndu) [inline osu.Game/Screens/Select/BeatmapCarousel.cs]: I would honestly consider carousel selecting the next beatmap when refining your filter a feature and wouldn't change it. I can see a similar scenario to what I laid out in #2638 happening if you refine your current filter while your currently selected map would get filtered out by your new filter.

Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs]: I don't follow.

Comment (LittleEndu) [reply osu.Game/Screens/Select/BeatmapCarousel.cs]: No. The selection happens in SelectNextRandom or in base.PerformSelection. This change (on this line at least) is making carousel select a random beatmap when there hasn't been a selection or when the last selection is filtered.

Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs]: Correct. That's the intention.

Comment (LittleEndu) [reply osu.Game/Screens/Select/BeatmapCarousel.cs]: I did, that's #2638. The change on this line is making the carousel select something away from beatmaps near previous selection and instead select a random one. "Fix carousel not returning to previous selection after filter" is literally the title of the PR, I fail to see how this is out of scope.

Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs]: Please wait until this fix is merged and then rebase your proposed change. I think you are misunderstanding this PR (please see the provided tests).

Comment (bdach) [reply osu.Game/Screens/Select/BeatmapCarousel.cs]: So I think @LittleEndu has an issue with behaviour tested in `RandomFallbackOnNonMatchingPrevious` - the reasoning is that when the previously-selected beatmap is filtered, the carousel should make a "local" change (as in choose from maps in proximity to the filtered ones) instead of making a random one. Whether that's the behaviour we want in the end is disputable and sort of up to preference, so I can't say anything objective about this. Subjectively I don't really care either way myself.

**Comment by bdach** [comment]
Don't have any issue with the code here, or any strong opinion on the discussion about desired behaviour.

File: osu.Game/Screens/Select/BeatmapCarousel.cs
Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs]: Yeah, I think I get that. Not sure it's correct though, and the case I mention in his new PR (where do you draw a line) is probably what we want to consider when deciding a path forward. ie. if 30000 beatmaps are hidden from a filter, do you just select the first remaining one? seems potentially worse than random. why even have random if it's only going to select once ever?

i dunno.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**
